### ISSUE TITLE
"Paths" in __NEXT_DATA__ are picked up by Google Search Console as internal links

### ISSUE THREAD

**Issue by tills13** [vercel/next.js#40143]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Wed Feb 19 06:37:35 UTC 2020
    Binaries:
      Node: 12.17.0
      npm: 6.14.4
      Yarn: 1.22.10
      pnpm: N/A
    Relevant packages:
      next: 12.2.0
      eslint-config-next: N/A
      react: 17.0.2
      react-dom: 17.0.2

### What browser are you using? (if relevant)

N/A

### How are you deploying your application? (if relevant)

N/A

### Describe the Bug

I recognize that this might not even be something NextJS can "fix" but I just want to highlight a problem we're seeing with our NextJS sites. I know there are people from Google who actively support NextJS, and they don't expose their own services to issues / feature requests, so I thought I might as well try here. 

tl;dr anything that even remotely resembles a path, returned as part of `__NEXT_DATA__`, will be picked up by Google Search Console and used as part of its crawling scope. 

The simplest `__NEXT_DATA__` is something that looks like

```
{ ... "page":"/docs/[[...slug]]" ...}
```

As seen here: https://nextjs.org/docs/basic-features/data-fetching/get-static-props. 

During its crawls, Google matches _within_ `__NEXT_DATA__`, seemingly anything that starts with a slash, and interprets them as internal links. In the Search Console, this manifests as a non-resolvable page, e.g. `https://nextjs.org/docs/[[...slug]`, with a referrer of `https://nextjs.org/docs/basic-features/data-fetching/get-static-props` (and any other page under `/docs`). 

**Note** that in the case of this NextJS page, it might not manifest as I've described as `/docs` is a valid page. For my particular use case, using a custom server, we're seeing the issue come up as `https://blog.com/post` with referrers of blog posts (e.g. `https://blog.com/the-post-slug`) -- `/post` is our _NextJS_ page (`pages/post.js`), but is otherwise not a valid URL. Let me know if you want some concrete examples, I've tried to make this as generic as possible. 

![image](https://user-images.githubusercontent.com/1269421/187771186-ab16389b-a753-4a63-9fc4-864f6ae2c32a.png)

![image](https://user-images.githubusercontent.com/1269421/187771267-856191e3-27f2-4fcc-9846-c8bfc219bd42.png)

**Why is this an issue?**

Internal 404s are interpreted as a site quality signal. Too many of them and Google figures your site is not being maintained and lowers your overall search ranking.

I _really_ don't want to have to implement dummy pages for `/post` and other slash-prefixed strings we have in `__NEXT_DATA__`

### Expected Behavior

Path like strings in `__NEXT_DATA__` should not be considered valid internal links by Google.  

### Link to reproduction

https://nextjs.org/docs/basic-features/data-fetching/get-static-props

### To Reproduce

1. build and deploy NextJS site
2. have site crawled by Google
3. note that Google finds and crawls internal NextJS pages, and anything else returned from data-fetching fns that looks like a path
4. these pages are marked as non-crawlable in Google Search Console

**Comment by andremendonca** [comment]
there is a discussion about this at https://github.com/vercel/next.js/discussions/39377 as well

**Comment by karoun** [comment]
For a potential fix inside Next, in similar circumstances, I recall that escaping the forward slashes (`"\/docs\/[[...slug]]"`) was sufficient to prevent Google from attempting to crawl the embedded URLs.

**Comment by p0zi** [comment]
> __NEXT_DATA__

Can you please provide a recipe how can i sscape with forward slashes?

**Comment by p0zi** [comment]
> > @karoun Can you please provide a recipe how can i sscape with forward slashes?
> 
> @p0zi not for Next, unfortunately. The workaround implemented was for a different framework, and involved doing `urls.replaceAll('/', '"')` on the server and `urls.replaceAll('"', '/')` on the client. That way, the contents of the HTML didn't include crawl-able URLs, but at JavaScript runtime the URLs were hydrated and correct.

Thank you very much for explanation!

**Comment by p0zi** [comment]
So far i added rel="nofollow" and will observe how Google crawler will behave.

`<script id="__NEXT_DATA__" type="application/json" rel="nofollow">{"props": ... }</script>`

**Comment by d-vorobyov** [comment]
@p0zi hi, how exactly did you add this to the next app?

**Comment by p0zi** [comment]
@d-vorobyov At the moment solution is dirty i modified module source code in file:
node_modules/next/dist/pages/_document.js in render() this is valid for Next.js v12

`
    render() {
        const { assetPrefix , inAmpMode , buildManifest , unstable_runtimeJS , docComponentsRendered , devOnlyCacheBusterQueryString , disableOptimizedLoading , crossOrigin ,  } = this.context;
        const disableRuntimeJS = unstable_runtimeJS === false;
        docComponentsRendered.NextScript = true;
        if (process.env.NEXT_RUNTIME !== "edge" && inAmpMode) {
            if (process.env.NODE_ENV === "production") {
                return null;
            }
            const ampDevFiles = [
                ...buildManifest.devFiles,
                ...buildManifest.polyfillFiles,
                ...buildManifest.ampDevFiles, 
            ];
            return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, disableRuntimeJS ? null : /*#__PURE__*/ _react.default.createElement("script", {
                id: "__NEXT_DATA__",
                type: "application/json",
                nonce: this.props.nonce,
                crossOrigin: this.props.crossOrigin || crossOrigin,
                dangerouslySetInnerHTML: {
                    __html: NextScript.getInlineScriptSource(this.context)
                },
                "data-ampdevmode": true
            }), ampDevFiles.map((file)=>/*#__PURE__*/ _react.default.createElement("script", {
                    key: file,
                    src: `${assetPrefix}/_next/${file}${devOnlyCacheBusterQueryString}`,
                    nonce: this.props.nonce,
                    crossOrigin: this.props.crossOrigin || crossOrigin,
                    "data-ampdevmode": true
                })));
        }
        if (process.env.NODE_ENV !== "production") {
            if (this.props.crossOrigin) console.warn("Warning: `NextScript` attribute `crossOrigin` is deprecated. https://nextjs.org/docs/messages/doc-crossorigin-deprecated");
        }
        const files = getDocumentFiles(this.context.buildManifest, this.context.__NEXT_DATA__.page, process.env.NEXT_RUNTIME !== "edge" && inAmpMode);
        return /*#__PURE__*/ _react.default.createElement(_react.default.Fragment, null, !disableRuntimeJS && buildManifest.devFiles ? buildManifest.devFiles.map((file)=>/*#__PURE__*/ _react.default.createElement("script", {
                key: file,
                src: `${assetPrefix}/_next/${encodeURI(file)}${devOnlyCacheBusterQueryString}`,
                nonce: this.props.nonce,
                crossOrigin: this.props.crossOrigin || crossOrigin
            })) : null, disableRuntimeJS ? null : /*#__PURE__*/ _react.default.createElement("script", {
            id: "__NEXT_DATA__",
            type: "application/json",
            nonce: this.props.nonce,
            crossOrigin: this.props.crossOrigin || crossOrigin,
            dangerouslySetInnerHTML: {
                __html: NextScript.getInlineScriptSource(this.context)
            },

rel: "nofollow"

        }), disableOptimizedLoading && !disableRuntimeJS && this.getPolyfillScripts(), disableOptimizedLoading && !disableRuntimeJS && this.getPreNextScripts(), disableOptimizedLoading && !disableRuntimeJS && this.getDynamicChunks(files), disableOptimizedLoading && !disableRuntimeJS && this.getScripts(files));
    }
`

**Comment by wieseljonas** [comment]
> So far i added rel="nofollow" and will observe how Google crawler will behave.
> 
> `<script id="__NEXT_DATA__" type="application/json" rel="nofollow">{"props": ... }</script>`

Any luck?

**Comment by p0zi** [comment]
Unfortunately it did not help.

But i am considering another aproach:
1) Backend side:
a) Data source with obfuscated link that will feed overwritten "next/link" component. Since i am using JSON:API from Drupal i can use path field enchancer to provide obfuscation with for example base_64 encoder preferably with some salt so Google cannot easily decode.
2) Frontend side:
a) Since our incoming data is already obfuscated we do not wory about __NEXT_DATA__.
b) No it is time to write our custom "next/link" component
tsconfig.json:
{
  "compilerOptions": {
    "paths": {
      "next/link": [
        "soft4net/components/Link/index.tsx"
      ]
    }
  }
}

`
import React from 'react';

// @see: https://stackoverflow.com/questions/68723485/how-to-setup-i18n-translated-url-routes-in-next-js
import { useRouter } from 'next/router';

// import Link from 'next/link'
import Link from 'node_modules/next/dist/client/link';


const ENCODE_SALT = `1qaz2wsx3edc4rfv5tgb`; // leverage if you need, this must match backend salt value, lenght for minimizing size of __NEXT_DATA___
/*
    @see: https://developer.mozilla.org/en-US/docs/Glossary/Base64#javascript_support
    JavaScript functions for decoding and encoding Base64:
    var encoded = btoa(stringValue);
    var decoded = atob(stringValue);
*/
export const urlEncode = (url) => {
    return btoa(url);
}
export const urlDecode = (urlEncoded) => {
    return atob(urlEncoded);
}

// @see: https://stackoverflow.com/questions/7860392/determine-if-string-is-in-base64-using-javascript
export function isBase64(str) {
    if (str === '' || str.trim() === '') {
        return false;
    }

    try {
        return btoa(atob(str)) == str;
    } catch (err) {
        return false;
    }
}

const CustomLink = ({ href, passHref = null, children, ...props }) => {
    const { locale } = useRouter();

    let _href = href;
    if (isBase64(_href)) {
        const hrefEncoded = _href;
        const hrefDecoded = urlDecode(hrefEncoded);
        _href = hrefDecoded;
    }

    return (
        <Link 
            {...props}
            href={_href} 
        >
            {children}
        </Link>
    )

}

export * from 'node_modules/next/dist/client/link';

export default CustomLink;
`

The idea is simple if path is obfuscated we know it should be decoded, we can also use this componenet for standard, non obfuscated paths.

This is a concept but i cannot see any issues with this concept.

**Comment by p0zi** [comment]
Unfortunately it did not help.
But i am considering another aproach:
1) Backend side:
a) Data source with obfuscated link that will feed overwritten "next/link" component. Since i am using JSON:API from Drupal i can use path field enchancer to provide obfuscation with for example base_64 encoder preferably with some salt so Google cannot easily decode.
2) Frontend side:
a) Since our ibncoming data is already obfuscated we do not woryy about __NEXT_DATA__.
b) No it is time to write our custom "next/link" component
tsconfig.json:
{
  "compilerOptions": {
    "paths": {
      "next/link": [
        "soft4net/components/Link/index.tsx"
      ]
    }
  }
}

`
import React from 'react'

// @see: https://stackoverflow.com/questions/68723485/how-to-setup-i18n-translated-url-routes-in-next-js
import { useRouter } from 'next/router'

// import Link from 'next/link'
import Link from 'node_modules/next/dist/client/link'


const ENCODE_SALT = '1qaz2wsx3edc4rfv5tgb'; // compropmise lenght for minimizing size of __NEXT_DATA___
/*
    @see: https://developer.mozilla.org/en-US/docs/Glossary/Base64#javascript_support
    JavaScript functions for decoding and encoding Base64:
    var encoded = btoa(stringValue);
    var decoded = atob(stringValue);
*/
export const urlEncode = (url) => {
    return btoa(url);
}
export const urlDecode = (urlEncoded) => {
    return atob(urlEncoded);
}

// @see: https://stackoverflow.com/questions/7860392/determine-if-string-is-in-base64-using-javascript
export function isBase64(str) {
    if (str === '' || str.trim() === '') {
        return false;
    }

    try {
        return btoa(atob(str)) == str;
    } catch (err) {
        return false;
    }
}

const CustomLink = ({ href, passHref = null, children, ...props }) => {
    const { locale } = useRouter();

    let _href = href;
    if (isBase64(_href)) {
        const hrefEncoded = _href;
        const hrefDecoded = urlDecode(hrefEncoded);
        _href = hrefDecoded;
    }

    return (
        <Link 
            {...props}
            href={_href} 
        >
            {children}
        </Link>
    )

}

export * from 'node_modules/next/dist/client/link';

export default CustomLink;

`

The idea is simple if path is obfuscated we know it should be decoded, we can also use this componenet for standard, non obfuscated paths.

**Comment by p0zi** [comment]
Unfortunately it did not help.
But i am considering another aproach:
1) Backend side:
a) Data source with obfuscated link that will feed overwritten "next/link" component. Since i am using JSON:API from Drupal i can use path field enchancer to provide obfuscation with for example base_64 encoder preferably with some salt so Google cannot easily decode.
2) Frontend side:
a) Since our ibncoming data is already obfuscated we do not woryy about __NEXT_DATA__.
b) No it is time to write our custom "next/link" component
tsconfig.json:
{
  "compilerOptions": {
    "paths": {
      "next/link": [
        "soft4net/components/Link/index.tsx"
      ]
    }
  }
}

`
import React from 'react'
import { useRouter } from 'next/router'
import Link from 'node_modules/next/dist/client/link' // import Link from 'next/link'


const ENCODE_SALT = '1qaz2wsx3edc4rfv5tgb'; // compropmise lenght for minimizing size of __NEXT_DATA___

export const urlEncode = (url) => {
    return btoa(url);
}

export const urlDecode = (urlEncoded) => {
    return atob(urlEncoded);
}

export function isBase64(str) {
    if (str === '' || str.trim() === '') {
        return false;
    }

    try {
        return btoa(atob(str)) == str;
    } catch (err) {
        return false;
    }
}

const CustomLink = ({ href, passHref = null, children, ...props }) => {
    const { locale } = useRouter();

    let _href = href;
    if (isBase64(_href)) {
        const hrefEncoded = _href;
        const hrefDecoded = urlDecode(hrefEncoded);
        _href = hrefDecoded;
    }

    return (
        <Link 
            {...props}
            href={_href} 
        >
            {children}
        </Link>
    )

}

export * from 'node_modules/next/dist/client/link';

export default CustomLink;
`

The idea is simple if path is obfuscated we know it should be decoded, we can also use this componenet for standard, non obfuscated paths.

**Comment by p0zi** [comment]
Unfortunately it did not help.
But i am considering another aproach:
1) Backend side:
a) Data source with obfuscated link that will feed overwritten "next/link" component. Since i am using JSON:API from Drupal i can use path field enchancer to provide obfuscation with for example base_64 encoder preferably with some salt so Google cannot easily decode.
2) Frontend side:
a) Since our ibncoming data is already obfuscated we do not woryy about __NEXT_DATA__.
b) No it is time to write our custom "next/link" component
tsconfig.json:
`
{
  "compilerOptions": {
    "paths": {
      "next/link": [
        "soft4net/components/Link/index.tsx"
      ]
    }
  }
}
`

`
import React from 'react'
import { useRouter } from 'next/router'
import Link from 'node_modules/next/dist/client/link' // import Link from 'next/link'


const ENCODE_SALT = '1qaz2wsx3edc4rfv5tgb'; // compropmise lenght for minimizing size of __NEXT_DATA___

export const urlEncode = (url) => {
    return btoa(url);
}

export const urlDecode = (urlEncoded) => {
    return atob(urlEncoded);
}

export function isBase64(str) {
    if (str === '' || str.trim() === '') {
        return false;
    }

    try {
        return btoa(atob(str)) == str;
    } catch (err) {
        return false;
    }
}

const CustomLink = ({ href, passHref = null, children, ...props }) => {
    const { locale } = useRouter();

    let _href = href;
    if (isBase64(_href)) {
        const hrefEncoded = _href;
        const hrefDecoded = urlDecode(hrefEncoded);
        _href = hrefDecoded;
    }

    return (
        <Link 
            {...props}
            href={_href} 
        >
            {children}
        </Link>
    )

}

export * from 'node_modules/next/dist/client/link';

export default CustomLink;

`

The idea is that simple if path is obfuscated we know it should be decoded, we can also use this componenet for standard, non obfuscated paths.

**Comment by p0zi** [comment]
Unfortunately it did not help.
But i am considering another aproach:
1) Backend side:
a) Data source with obfuscated link that will feed overwritten "next/link" component. Since i am using JSON:API from Drupal i can use path field enchancer to provide obfuscation with for example base_64 encoder preferably with some salt so Google cannot easily decode.
2) Frontend side:
a) Since our ibncoming data is already obfuscated we do not woryy about __NEXT_DATA__.
b) No it is time to write our custom "next/link" component
tsconfig.json:
`
{
  "compilerOptions": {
    "paths": {
      "next/link": [
        "soft4net/components/Link/index.tsx"
      ]
    }
  }
}
`

`
import React from 'react'
import { useRouter } from 'next/router'
import Link from 'node_modules/next/dist/client/link' // import Link from 'next/link'


const ENCODE_SALT = '1qaz2wsx3edc4rfv5tgb'; // compropmise lenght for minimizing size of __NEXT_DATA___

export const urlEncode = (url) => {
    return btoa(url);
}

export const urlDecode = (urlEncoded) => {
    return atob(urlEncoded);
}

export function isBase64(str) {
    if (str === '' || str.trim() === '') {
        return false;
    }

    try {
        return btoa(atob(str)) == str;
    } catch (err) {
        return false;
    }
}

const CustomLink = ({ href, passHref = null, children, ...props }) => {
    const { locale } = useRouter();

    let _href = href;
    if (isBase64(_href)) {
        const hrefEncoded = _href;
        const hrefDecoded = urlDecode(hrefEncoded);
        _href = hrefDecoded;
    }

    return (
        <Link 
            {...props}
            href={_href} 
        >
            {children}
        </Link>
    )

}

export * from 'node_modules/next/dist/client/link';

export default CustomLink;
`

The idea is that simple if path is obfuscated we know it should be decoded, we can also use this componenet for standard, non obfuscated paths.

**Comment by p0zi** [comment]
Unfortunately it did not help.
But i am considering another aproach:
1) Backend side:
a) Data source with obfuscated link that will feed overwritten "next/link" component. Since i am using JSON:API from Drupal i can use path field enchancer to provide obfuscation with for example base_64 encoder preferably with some salt so Google cannot easily decode.
2) Frontend side:
a) Since our ibncoming data is already obfuscated we do not woryy about __NEXT_DATA__.
b) No it is time to write our custom "next/link" component
tsconfig.json:
`
{
  "compilerOptions": {
    "paths": {
      "next/link": [
        "soft4net/components/Link/index.tsx"
      ]
    }
  }
}
`

---
import React from 'react'
import { useRouter } from 'next/router'
import Link from 'node_modules/next/dist/client/link' // import Link from 'next/link'


const ENCODE_SALT = '1qaz2wsx3edc4rfv5tgb'; // compropmise lenght for minimizing size of __NEXT_DATA___

export const urlEncode = (url) => {
    return btoa(url);
}

export const urlDecode = (urlEncoded) => {
    return atob(urlEncoded);
}

export function isBase64(str) {
    if (str === '' || str.trim() === '') {
        return false;
    }

    try {
        return btoa(atob(str)) == str;
    } catch (err) {
        return false;
    }
}

const CustomLink = ({ href, passHref = null, children, ...props }) => {
    const { locale } = useRouter();

    let _href = href;
    if (isBase64(_href)) {
        const hrefEncoded = _href;
        const hrefDecoded = urlDecode(hrefEncoded);
        _href = hrefDecoded;
    }

    return (
        <Link 
            {...props}
            href={_href} 
        >
            {children}
        </Link>
    )

}

export * from 'node_modules/next/dist/client/link';

export default CustomLink;
---


The idea is that simple if path is obfuscated we know it should be decoded, we can also use this componenet for standard, non obfuscated paths.

**Comment by p0zi** [comment]
Unfortunately it did not help.
But i am considering another aproach:
1) Backend side:
a) Data source with obfuscated link that will feed overwritten "next/link" component. Since i am using JSON:API from Drupal i can use path field enchancer to provide obfuscation with for example base_64 encoder preferably with some salt so Google cannot easily decode.
2) Frontend side:
a) Since our ibncoming data is already obfuscated we do not woryy about __NEXT_DATA__.
b) No it is time to write our custom "next/link" component
tsconfig.json:
`
	{
	  "compilerOptions": {
	    "paths": {
	      "next/link": [
		"soft4net/components/Link/index.tsx"
	      ]
	    }
	  }
	}
`

soft4net/components/Link/index.tsx:
`
	import React from 'react'
	import { useRouter } from 'next/router'
	import Link from 'node_modules/next/dist/client/link' // import Link from 'next/link'


	const ENCODE_SALT = '1qaz2wsx3edc4rfv5tgb'; // compropmise lenght for minimizing size of __NEXT_DATA___

	export const urlEncode = (url) => {
	    return btoa(url);
	}

	export const urlDecode = (urlEncoded) => {
	    return atob(urlEncoded);
	}

	export function isBase64(str) {
	    if (str === '' || str.trim() === '') {
		return false;
	    }

	    try {
		return btoa(atob(str)) == str;
	    } catch (err) {
		return false;
	    }
	}

	const CustomLink = ({ href, passHref = null, children, ...props }) => {
	    const { locale } = useRouter();

	    let _href = href;
	    if (isBase64(_href)) {
		const hrefEncoded = _href;
		const hrefDecoded = urlDecode(hrefEncoded);
		_href = hrefDecoded;
	    }

	    return (
		<Link 
		    {...props}
		    href={_href} 
		>
		    {children}
		</Link>
	    )

	}

	export * from 'node_modules/next/dist/client/link';

	export default CustomLink;
`

The idea is that simple if path is obfuscated we know it should be decoded, we can also use this componenet for standard, non obfuscated paths.

**Comment by muslu** [comment]
any news?

**Comment by falahati** [comment]
the whole `__NEXT_DATA__` should be obfuscated IMO. Google has already picked up information on the redux state on our site and translation phrases for parts of the website that are not available before logging in.

 I know I can do that on the server by modifying the `getInlineScriptSource`, but is there a similar function that is used to read the content? it seem to happen on the `client/index.tsx` file, not something we can easily modify except by patching the library itself.

**Comment by falahati** [comment]
on further tests, the following might solve your problem:
```typescript
import Document, { Head, Html, Main, NextScript } from "next/document";
import React from "react";

export default class MyDocument extends Document {
    public render(): JSX.Element {
        return <Html>
            <Head />
            <body>
                <Main />
                <NextScript />
                <script
                    lang="javascript"
                    type="text/javascript"
                    defer={false}
                    async={false}
                    dangerouslySetInnerHTML={
                        {
                            __html: `
                                const element = document.getElementById("__NEXT_DATA__");
                                element.innerText = atob(element.innerText);
                            `,
                        }
                    }
                />
            </body>
        </Html>;
    }
}

// eslint-disable-next-line @typescript-eslint/unbound-method
const nextInlineScriptSource = NextScript.getInlineScriptSource;
NextScript.getInlineScriptSource = (props) => {
    const value = nextInlineScriptSource(props);
    return Buffer.from(value).toString("base64");
};
```

put this in your `_document.tsx` file

**Comment by c0b41** [comment]
@falahati any workaround for app router

**Comment by utkusezici** [comment]
![githubimage](https://github.com/vercel/next.js/assets/36607135/faf7e820-09ab-49b3-b5e4-a88cc3a3a43f)
@falahati, I want to fix this issue, but I'm getting a Hydration error when adding the code you provided.

**Comment by falahati** [comment]
what is the error? the `nextInlineScriptSource` is only used for the content of the `__next_data__` tag from what I see in the source code and the `<script>` should be executed before any next related javascript. so I don't know how it got picked with the hydration code before it got executed if that is in fact the issue.

could it be that you were not using a `_document.tsx` file and somehow its mere presence broke your code? what happens if you remove the `nextInlineScriptSource` mutation and the script tag but keep the `_document.tsx` file intact?

full disclosure, I am on a fairly old version (v12) of next in this project and we are not using any of the new features. for example, app router was added in v13 I think. however, I won't see how it could break your code still. at least from the source code, I don't see any other place being dependent on the `nextInlineScriptSource` method.

**Comment by falahati** [comment]
unless app router uses the same `nextInlineScriptSource` method somewhere else. in that case, we have to find a way to only obfuscate the first response. now I don't really know how it works and I have not used app router. but I am sure checking the network activity of the application could help you find the reason. in any case, I hope this piece of code if not useful in its entirety could at least be used to point to some sort of final workaround.

**Comment by falahati** [comment]
if anyone tried to use the code above be advised that `atob` is not friendly with Unicode chars. I had a few problems with it.

**Comment by Fi1osof** [comment]
> __NEXT_DATA__

How you add rel nofollow?

**Comment by falahati** [comment]
I have personally changed the client side script to this solving the issue:
```javascript
const element = document.getElementById("__NEXT_DATA__");
const data = atob(element.innerText);
const bytes = new Uint8Array(data.length);
for (let b = 0; b < bytes.length; ++b) {
   bytes[b] = data.charCodeAt(b);
}
element.innerText = new TextDecoder('utf-8').decode(bytes);
```
the server-side code doesn't need to be changed. `Buffer` already encodes the data in binary. this will convert back the decoded ASCII string into binary and parse it as utf8 binary, converting it to utf8 string.

**Comment by za1x** [comment]
any info?

**Comment by Eugene0** [comment]
Hi @falahati Thanks for sharing your solution with obfuscating `__NEXT_DATA__` via `getInlineScriptSource`. I’ve tested it in my project, and it works great.

However, I have one concern.
If the code inside `getInlineScriptSource` runs correctly, but the `<script>` inside the `MyDocument class` fails to execute, I end up with a completely broken site — just a blank white screen. When I manually remove the white overlay, I can see that hydration is broken and some styles are missing. Right now, I can only reproduce this by commenting out the `<script> `logic manually. But if this could ever happen in real-world scenarios — even rarely — it would be a very critical issue for me.

From what I understand, you’ve been using this approach for about a year now.
Have you ever encountered a case where the `<script>` inside `MyDocument` didn’t run correctly?
Or have you seen any other issues arise as a result of obfuscating `__NEXT_DATA__`?

**Comment by falahati** [comment]
Yes of course, if it dont get executed, you get the application error since NextJS fails to understand the state in the obscured form. This piece of code, as described in my 12 Jul 2024 comment, is still running strong on the production environment it was first introduced in, and I had no problem with it or any report of such issues.

If the script for some reason dont get executed in time, it means that the browser has a broken policy for running scripts, which I dont think is something we have to worry about. In my comment, I have put it right after the next state script, and also prevented it from being deferred or executed as async, so it should essentially block the browser thread, preventing any interference by other pieces of code in the page, including NextJS internal code; short of the browser ignoring the specific attributes attribues attached to the script tag for some reason, if the script is placed in the right place and with the right attributes, I can not see a reason for it failing.

**Comment by TvickyPace** [comment]
Guys, its 2026, is still no proper solution? facing same issue, i have around 50k pages, and they all give same issue of google crawlling links in script. let me know if any of you find good approch to it.

**Comment by muslu** [comment]
any news?
